### PR TITLE
CATTY-320 Implement SetPenColorBrick

### PR DESCRIPTION
**Pull request by neel-makhecha**
- Implemented SetPenColorBrick, its instructions and serialization with tests.
- Implemented SetPenColorBrickCell

### Your checklist for this pull request
Please review the [contributing guidelines](https://github.com/Catrobat/Catty/blob/develop/README.md) and [wiki pages](https://github.com/Catrobat/Catroid/wiki/) of this repository.

- [ ] Include the name of the Jira ticket in the PR’s title
- [ ] Verify that the Jira ticket is in the status *Ready for Development*
- [ ] Include a summary of the changes plus the relevant context
- [ ] Choose the proper base branch (*develop*)
- [ ] Confirm that the changes follow the project’s coding guidelines
- [ ] Verify that the changes generate no compiler or linter warnings
- [ ] Perform a self-review of the changes
- [ ] Verify to commit no other files than the intentionally changed ones
- [ ] Include reasonable and readable tests verifying the added or changed behavior
- [ ] Confirm that new and existing unit tests pass locally
- [ ] Check that the commits’ message style matches the [project’s guideline](https://github.com/Catrobat/Catroid/wiki/Commit-Message-Guidelines)
- [ ] Stick to the project’s git workflow (rebase and squash your commits)
- [ ] Verify that your changes do not have any conflicts with the base branch
- [ ] After the PR, verify that all CI checks have passed
- [ ] Post a message in the *#catty* [Slack channel](https://catrobat.slack.com) and ask for a code reviewer

### CODE REVIEW

File: src/Catty/Extension&Delegate&Protocol/Extensions/UIColorExtension.swift [hunk 1]
```diff
@@ -36,6 +36,19 @@ extension UIColor {
     /*
      * from: https://stackoverflow.com/a/24263296
      */
+
+    var redComponent: Int {
+        Int(self.cgColor.components![0] * 255)
```
Comment (m-herold) [inline src/Catty/Extension&Delegate&Protocol/Extensions/UIColorExtension.swift:41]: Please avoid force unwrapping in production code and add a test for each component (`CattyTests/Extensions/UIColorEXtensionTests`).

File: src/CattyTests/Bricks/SetPenColorBrickTests.swift [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+ *  along with this program.  If not, see http://www.gnu.org/licenses/.
+ */
+
+import XCTest
+
+@testable import Pocket_Code
+
+final class SetPenColorBrickTests: AbstractBrickTest {
+
+    func testSetPenColorBrick() {
```
Comment (m-herold) [inline src/CattyTests/Bricks/SetPenColorBrickTests.swift:29]: Please add additional tests verifying what is happening when this Brick is executed with negative values, string formulas, values greater than INT_MAX, or smaller than INT_MIN (e.g., `testSetPenColorBrickWithFormulaLargerThanIntMax`, `testSetPenColorBrickWithNegativeFormula`, `testSetPenColorBrickWithStringFormula`)

File: src/CattyTests/Parser & Serialization/CatrobatLanguage0.991/XMLSerializerTests/Unit Tests/XMLSerializerTests.swift [hunk 3]
```diff
@@ -145,4 +145,19 @@ final class XMLSerializerTests: XMLAbstractTest {
...
         XCTAssertTrue(equal, "XMLElement invalid!")
     }
+
+    func testSetPenColorBrick() {
+        let project = self.getProjectForXML(xmlFile: "ValidProjectAllBricks0991")
+
+        let brick = ((project.objectList.object(at: 0) as! SpriteObject).scriptList.object(at: 0) as! Script).brickList.object(at: 45) as! SetPenColorBrick
+        let xmlElementPath = "//program/objectList/object[1]/scriptList/script[1]/brickList/brick[46]"
+
+        guard let xmlElement = brick.xmlElement(with: CBXMLSerializerContext()) else {
```
Comment (m-herold) [inline src/CattyTests/Parser & Serialization/CatrobatLanguage0.991/XMLSerializerTests/Unit Tests/XMLSerializerTests.swift:155]: Please rebase from the current version of develop and fix this test.

File: src/CattyTests/Bricks/SetPenColorBrickTests.swift (hunk 2 above)
Comment (neel-makhecha) [reply src/CattyTests/Bricks/SetPenColorBrickTests.swift:29]: In the UIColor extension, the convenience init was previously checking whether the values are in range using assert statements. So to make it consistent with the behaviour on Catroid, I have removed those statements, as UIColor handles it internally and it is consistent with Catroid as well. That also includes negative values. For string values, the formula editor will give a syntax error so that isn't a possibility. Please let me know if I am missing out anything.

File: src/CattyTests/Extensions/UIColorExtensionTests.swift [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+ *  along with this program.  If not, see http://www.gnu.org/licenses/.
+ */
+
+import XCTest
+
+@testable import Pocket_Code
+
+class UIColorExtensionTests: XCTestCase {
+
+    func testRGBComponentProperties() {
```
Comment (m-herold) [inline src/CattyTests/Extensions/UIColorExtensionTests.swift:29]: Please also document and test what is happening with that extension in case of negative or out-of-range values (e.g., larger than 255).

File: src/CattyTests/Parser & Serialization/CatrobatLanguage0.991/XMLSerializerTests/Unit Tests/XMLSerializerTests.swift (hunk 3 above)
Comment (m-herold) [reply src/CattyTests/Parser & Serialization/CatrobatLanguage0.991/XMLSerializerTests/Unit Tests/XMLSerializerTests.swift:159]: The initializer of `CBXMLSerializerContext` has changed, that's probably the reason why your build is not compiling.

File: src/CattyTests/Bricks/SetPenColorBrickTests.swift (hunk 2 above)
Comment (m-herold) [reply src/CattyTests/Bricks/SetPenColorBrickTests.swift:29]: You are using `interpretInteger` in your implementation which is very clever. Other Bricks do not do that and manually cast the result (double) to integer which of course  is crashing the app for out-of-range values. That was one of the reason we needed to make a hotfix release last week. So please verify that your Brick also works with those two out-of-range values (greater `INT_MAX` and smaller `INT_MIN`) -- it is a necessary documentation in case someone removes or refactors the `interpretInteger` or your Brick instruction.

I have a feeling that the app crashes with your Brick having a Formula evaluating to smaller than `INT_MIN`. Thanks a lot for pointing that out -- that probably just saved us a lot of time debugging that bug (ideally it would be awesome if you can also add `testInterpretIntegerMin` to `FormulaManagerInterpreterTests` then you know what I am talking about).
